### ISSUE TITLE
Error with _fblas

### ISSUE THREAD

**Issue by zoeherri** [scipy/scipy#11826]
I am trying to use scipy.spatial.distance.cdist
I can import scipy but cannot call or import scipy.spatial
I have searched the web and found other posts similar to this but none have worked for me. I tried different versions of scipy, numpy and python itself. I continue to get the error shown in the image below. I do not appear to have _fblas.py. I searched the C drive of my computer and all it returned was test_fblas.py, which is present in the test folder within the scipy folder. I appreciate any help leading to the resolution of this hours long quest.
![error](https://user-images.githubusercontent.com/63371401/78828254-6c84bf00-79b2-11ea-9e1e-954a3ae7601e.jpg)

**Comment by petacreepers23** [comment]
I have the same issue, couldn't find a fix either

**Comment by rgommers** [comment]
Could you please add some details on how you installed Python, how you installed SciPy, and provide relevant debugging info like `$ ls site-packages/scipy/linalg/` and the output of `scipy.show_config()`?

**Comment by petacreepers23** [comment]
i am on windows, installed python via windows 10 store.
scipy installed via pip using command shown in https://www.scipy.org/install.html

the ls you ask for:
```
hon37\site-packages\scipy\linalg> ls


    Directorio: C:\Users\reiva\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\scipy\linalg


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       13/04/2020     18:04                src
d-----       13/04/2020     18:04                tests
d-----       13/04/2020     18:04                __pycache__
-a----       13/04/2020     18:04          56481 basic.py
-a----       13/04/2020     18:04          10519 blas.py
-a----       13/04/2020     18:04         211456 cython_blas.cp37-win_amd64.pyd
-a----       13/04/2020     18:04          14717 cython_blas.pxd
-a----       13/04/2020     18:04         644608 cython_lapack.cp37-win_amd64.pyd
-a----       13/04/2020     18:04         195600 cython_lapack.pxd
-a----       13/04/2020     18:04          52862 decomp.py
-a----       13/04/2020     18:04          11790 decomp_cholesky.py
-a----       13/04/2020     18:04           6757 decomp_lu.py
-a----       13/04/2020     18:04          13497 decomp_qr.py
-a----       13/04/2020     18:04          10318 decomp_schur.py
-a----       13/04/2020     18:04          14706 decomp_svd.py
-a----       13/04/2020     18:04           1728 flinalg.py
-a----       13/04/2020     18:04          30964 interpolative.py
-a----       13/04/2020     18:04          12944 lapack.py
-a----       13/04/2020     18:04            159 linalg_version.py
-a----       13/04/2020     18:04          18280 matfuncs.py
-a----       13/04/2020     18:04           6453 misc.py
-a----       13/04/2020     18:04           6058 setup.py
-a----       13/04/2020     18:04          34481 special_matrices.py
-a----       13/04/2020     18:04          12102 _cython_signature_generator.py
-a----       13/04/2020     18:04          12575 _decomp_ldl.py
-a----       13/04/2020     18:04           3623 _decomp_polar.py
-a----       13/04/2020     18:04          14611 _decomp_qz.py
-a----       13/04/2020     18:04         251392 _decomp_update.cp37-win_amd64.pyd
-a----       13/04/2020     18:04          12387 _expm_frechet.py
-a----       13/04/2020     18:04         580608 _fblas.cp37-win_amd64.pyd
-a----       13/04/2020     18:04        1467392 _flapack.cp37-win_amd64.pyd
-a----       13/04/2020     18:04          51200 _flinalg.cp37-win_amd64.pyd
-a----       13/04/2020     18:04          25132 _generate_pyx.py
-a----       13/04/2020     18:04         225792 _interpolative.cp37-win_amd64.pyd
-a----       13/04/2020     18:04          45192 _interpolative_backend.py
-a----       13/04/2020     18:04          28050 _matfuncs_inv_ssq.py
-a----       13/04/2020     18:04           5874 _matfuncs_sqrtm.py
-a----       13/04/2020     18:04           2787 _procrustes.py
-a----       13/04/2020     18:04           5920 _sketches.py
-a----       13/04/2020     18:04          28279 _solvers.py
-a----       13/04/2020     18:04         192512 _solve_toeplitz.cp37-win_amd64.pyd
-a----       13/04/2020     18:04           1814 _testutils.py
-a----       13/04/2020     18:04           7406 __init__.py

```



the output of the config:

```
>>> scipy.show_config()
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    library_dirs = ['C:\\projects\\scipy-wheels\\scipy\\build\\openblas']
    libraries = ['openblas']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    library_dirs = ['C:\\projects\\scipy-wheels\\scipy\\build\\openblas']
    libraries = ['openblas']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    library_dirs = ['C:\\projects\\scipy-wheels\\scipy\\build\\openblas']
    libraries = ['openblas']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    library_dirs = ['C:\\projects\\scipy-wheels\\scipy\\build\\openblas']
    libraries = ['openblas']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]

```





also, apart from 
`from scipy import linalg`
this
`from scipy import signal`
also fails to load the _fblas

**Comment by rgommers** [comment]
There's a file `_fblas.cp37-win_amd64.pyd` right there, it's unclear to me why that wouldn't be importable. Can you import it if you `cd` to that directory and directly do `import _fblas`?

> installed python via windows 10 store.

unless there's a recurrence of the DLL load library path issue here ....

@zoeherri did you also install Python from the Windows Store?

**Comment by rgommers** [comment]
> There's a file `_fblas.cp37-win_amd64.pyd` right there, it's unclear to me why that wouldn't be importable. Can you import it if you `cd` to that directory in your Python interpreter and directly do `import _fblas`?
> 
> > installed python via windows 10 store.
> 
> unless there's a recurrence of the DLL load library path issue here ....
> 
> @zoeherri did you also install Python from the Windows Store?

**Comment by petacreepers23** [comment]
@rgommers didn't work, same error "cannot find module"

```
PS C:\Users\reiva\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\scipy\linalg> python
Python 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 11:52:54) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import _fblas
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: DLL load failed: No se puede encontrar el módulo especificado.
```

**Comment by rgommers** [comment]
Yes, I think this is a duplicate of gh-10171, Windows Store Python 3.7 is problematic. Could you upgrade to 3.8?

**Comment by rgommers** [comment]
Oh this brings back bad memories. @tylerjereddy the `distributor_init.py` is only copied into our wheels for Python 3.8: https://github.com/MacPython/scipy-wheels/blob/9da5817775eedebfa0eb929ae6a007085ad8d60b/appveyor.yml#L176

```
Python 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020
```
This is a new micro release. I suspect the DLL path resolution was changed in that release.

@petacreepers23 could you download https://github.com/MacPython/scipy-wheels/blob/master/_distributor_init.py and put it in the main `scipy` directory, and see if that fixes things for you when you do `from scipy import linalg` (or any other module that fails right now)?

**Comment by petacreepers23** [comment]
ok @rgommers i uninstalled windows store version, and installed python 3.7.7 from python.org, I cannot use python 3.8 because of other dependencies.
removed all win_store envioroment variables, replacing them with the new python install and upgrade pip and reinstall all, and now it works with no problem, without downloading that.

It seems the problem is microsoft store python... too bad for that UWP api that might use...

**Comment by rgommers** [comment]
Thanks Tyler. Let's see if more reports come in and if we get confirmation that the "copy in `_distributor_init.py` file fixes it. Still a bit fuzzy on whether this really came in with the new Python 3.7.7 or not.

**Comment by petacreepers23** [comment]
I found some weird behaviour, not affecting my development of apps, but just as a curiosity, 
i did cd to `site-packages/scipy/linalg/` just like at the begining, opened python there and tried `import _fblas` and the error persists, but, importing linalg from the whole lib don't fail:

```
PS C:\Users\reiva\AppData\Local\Programs\Python\Python37\Lib\site-packages\scipy\linalg> python
Python 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 10:41:24) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import _fblas
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: DLL load failed: No se puede encontrar el módulo especificado.
>>> from scipy import linalg
>>> from scipy.linalg import _fblas
```
Also the _distributor_init.py is in this normal version of python, but the file just have some commented lines. I'll try the win10store version and let you know.

**Comment by rgommers** [comment]
> Also the _distributor_init.py is in this normal version of python, but the file just have some commented lines. I'll try the win10store version and let you know.

Yeah there's a default dummy file, which needs to be replaced depending on how binaries are built.

**Comment by petacreepers23** [comment]
ok, reinstalled win10 python, and still have problems while importing linalg.
I tried replacing with 

> @petacreepers23 could you download https://github.com/MacPython/scipy-wheels/blob/master/_distributor_init.py and put it in the main scipy directory, and see if that fixes

and it works! , with that file i can import signal and linalg

**Comment by rgommers** [comment]
Great thanks @petacreepers23. That's at least an easy workaround for others who hit this issue. We can include this fix in the next release.

**Comment by tylerjereddy** [comment]
I have a question here since I've been struggling a bit with the fix in our wheels repo--would it be "ok" to only apply the fix to 64-bit Windows wheels with Python 3.7? I noticed in the Windows store that Python 3.7 has "x64" as minimum arch requirement, so technically users should have the option to use 64-bit mode?

https://www.microsoft.com/en-us/p/python-37/9nj46sx7x90p?activetab=pivot:regionofsystemrequirementstab

Another perhaps-reasonable excuse for this shortcut is that Python 3.7 in windows store is described as experimental, so that 32-bit DLL resolution rabbit hole might best be revisited later:

>NOTE: The Microsoft Store release of Python 3.7 is primarily for evaluation purposes, and not all features are guaranteed stable. See https://docs.python.org/3.7/using/windows.html for more information, or consider searching for Python 3.8 in the store instead.

**Comment by rgommers** [comment]
So is adding the `_distributor_init.py` breaking something?

Leaving out 32-bit for now seems okay if needed. I would make sure it's unconditionally applied for 64-bit though, otherwise we'll run into it again for 3.6 or 3.9

**Comment by zoeherri** [comment]
> There's a file `_fblas.cp37-win_amd64.pyd` right there, it's unclear to me why that wouldn't be importable. Can you import it if you `cd` to that directory and directly do `import _fblas`?
> 
> > installed python via windows 10 store.
> 
> unless there's a recurrence of the DLL load library path issue here ....
> 
> @zoeherri did you also install Python from the Windows Store?

I did not install from the windows store. I used https://www.python.org/downloads/ to get the latest version and to pull 3.7 in case it was a but with the latest. I used pip to uninstall and reinstall scipy but when reinstalling it calls from a cash. I am currently uninstalling python and all related libraries to start from scratch.
I have another computer which had never had python on it before so I installed the latest version from the site above, used pip to install numpy and scipy then tried and had no errors. I initially had jupyter notebook and anaconda on the machine having problems so I wonder if this is causing an issue

**Comment by zoeherri** [comment]
I ended up clearing the cache, reinstalling and then telling it to update scipy once installed. I can now import scipy and associated libraries.

**Comment by tylerjereddy** [comment]
I think I'm reasonably happy with the current version of the proposed fix in: https://github.com/MacPython/scipy-wheels/pull/70

It adds a `_distributor_init.py` to all versions of Python wheel we support for Windows (32 and 64-bit/Python 3.6, 3.7, 3.8) in next release.

The main trick was to add a `os.chdir()` call to move right into the DLL path, which seems to help with DLL resolution for older Python versions; it should have no effect for Python 3.8+. It moves back to the original dir after.

The only remaining failures in that PR are unrelated: https://github.com/scipy/scipy/issues/11095

**Comment by rgommers** [comment]
The fix was merged, so let's close this. The 1.5.0 wheels will have the fix, if anyone runs into this with 1.4.1 please use this workaround: https://github.com/scipy/scipy/issues/11826#issuecomment-613052040 (a 1.4.2 may follow, but not guaranteed).

Thanks everyone!
